src/Microsoft.Azure.ServiceBus.KeyVault/KeyVaultPlugin.cs

```diff
@@ -0,0 +1,166 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Microsoft.Azure.ServiceBus.KeyVault
+{
+    using System;
+    using System.Threading.Tasks;
+    using Microsoft.Azure.ServiceBus.Core;
+    using System.Security.Cryptography;
+    using System.IO;
+
+    /// <summary>
+    /// Provides Azure KeyVault functionality for Azure Service Bus.
+    /// </summary>
+    public class KeyVaultPlugin : ServiceBusPlugin
+    {
+        private readonly string secretName;
+        private readonly string keyVaultEndpoint;
+        private KeyVaultSecretManager secretManager;
+        private byte[] initializationVector;
+        private string base64InitializationVector;
+
+        /// <summary>
+        /// Gets the name that is used to identify this plugin.
+        /// </summary>
+        public override string Name => "Microsoft.Azure.ServiceBus.KeyVault.KeyVaultPlugin";
+
+        /// <summary>
+        /// Creates a new instance of an <see cref="KeyVaultPlugin"/>.
+        /// </summary>
+        /// <param name="encryptionSecretName">The name of the secret used to encrypt / decrypt messages.</param>
+        /// <param name="options">The <see cref="KeyVaultPluginOptions"/> used to create a new instance.</param>
+        public KeyVaultPlugin(string encryptionSecretName, KeyVaultPluginOptions options)
+        {
+            if (string.IsNullOrEmpty(encryptionSecretName))
+            {
+                throw new ArgumentNullException(nameof(encryptionSecretName));
+            }
+            if (options == null)
+            {
+                throw new ArgumentNullException(nameof(options));
+            }
+
+            this.secretName = encryptionSecretName;
+            this.keyVaultEndpoint = options.Endpoint;
+            this.secretManager = new KeyVaultSecretManager(options.Endpoint, options.ClientId, options.ClientSecret);
+            this.initializationVector = KeyVaultPlugin.GenerateInitializationVector();
+            this.base64InitializationVector = Convert.ToBase64String(this.initializationVector);
+        }
+
+        /// <summary>
+        /// The action performed before sending a message to Service Bus. This method will load the KeyVault key and encrypt messages.
+        /// </summary>
+        /// <param name="message">The <see cref="Message"/> to be encrypted.</param>
+        /// <returns>The encrypted <see cref="Message"/>.</returns>
+        public override async Task<Message> BeforeMessageSend(Message message)
+        {
+            try
+            {
+                if (message.UserProperties.ContainsKey(KeyVaultMessageHeaders.InitializationVectorPropertyName) || message.UserProperties.ContainsKey(KeyVaultMessageHeaders.KeyNamePropertyName))
```

Comment:
Yeah, but that is a not, so if either are missing, it will just return the message...